pkg/controller/tas/node_failure_controller.go

```diff
@@ -185,57 +201,58 @@ func (r *nodeFailureReconciler) patchWorkloadsForNodeToReplace(ctx context.Conte
 	}
 	var workloadProcessingErrors []error
 	for wlKey := range workloadsToProcess {
+		log := r.log.WithValues("workload", wlKey, "nodeName", nodeName)
+		// fetch workload.
 		var wl kueue.Workload
 		if err := r.client.Get(ctx, wlKey, &wl); err != nil {
 			if apierrors.IsNotFound(err) {
-				r.log.V(4).Info("Workload not found, skipping", "workload", wlKey)
+				log.V(4).Info("Workload not found, skipping")
 			} else {
-				r.log.V(2).Error(err, "Failed to get workload", "workload", wlKey)
+				log.V(2).Error(err, "Failed to get workload")
 				workloadProcessingErrors = append(workloadProcessingErrors, err)
 			}
 			continue
 		}
-		annotations := getAnnotations(&wl)
-		evictedNow := false
-		if failedNode, ok := annotations[kueuealpha.NodeToReplaceAnnotation]; ok && failedNode != nodeName && !workload.IsEvicted(&wl) {
-			r.log.V(3).Info("Evicting workload due to multiple node failures", "workload", wlKey)
-			evictionMsg := fmt.Sprintf(NodeMultipleFailuresEvictionMessageFormat, failedNode, nodeName)
-			if evictionErr := r.startEviction(ctx, &wl, evictionMsg); evictionErr != nil {
-				r.log.V(2).Error(evictionErr, "Failed to complete eviction process", "workload", wlKey)
-				workloadProcessingErrors = append(workloadProcessingErrors, evictionErr)
-				continue
-			} else {
-				evictedNow = true
-			}
+
+		// evict workload when annotation present.
+		evicted, err := r.evictWorkload(ctx, log, &wl, wlKey, nodeName)
+		if err != nil {
+			workloadProcessingErrors = append(workloadProcessingErrors, err)
+			continue
+		}
+		// re-fetch workload if we evicted it.
+		if evicted {
 			if err := r.client.Get(ctx, wlKey, &wl); err != nil {
-				r.log.V(2).Error(err, "Failed to re-fetch workload after eviction", "workload", wlKey)
+				log.V(2).Error(err, "Failed to re-fetch workload after eviction")
 				workloadProcessingErrors = append(workloadProcessingErrors, err)
 				continue
 			}
 		}
-		err := clientutil.Patch(ctx, r.client, &wl, true, func() (bool, error) {
-			annotations = getAnnotations(&wl)
+
+		// update annotations.
+		err = clientutil.Patch(ctx, r.client, &wl, true, func() (bool, error) {
+			annotations := getAnnotations(&wl)
 			failedNode, ok := annotations[kueuealpha.NodeToReplaceAnnotation]
 			if !ok {
-				r.log.V(4).Info(fmt.Sprintf("Adding node to %s annotation", kueuealpha.NodeToReplaceAnnotation), "workload", wlKey, "nodeName", failedNode)
+				log.V(4).Info(fmt.Sprintf("Adding node to %s annotation", kueuealpha.NodeToReplaceAnnotation))
 				annotations[kueuealpha.NodeToReplaceAnnotation] = nodeName
 				wl.SetAnnotations(annotations)
 				return true, nil
 			}
-			if workload.IsEvicted(&wl) || evictedNow {
-				r.log.V(4).Info(fmt.Sprintf("Removing node from %s annotation", kueuealpha.NodeToReplaceAnnotation), "workload", wlKey, "nodeName", failedNode)
+			if workload.IsEvicted(&wl) || evicted {
```

Comment:
feel free to keep as is if you disagree, but I liked the name evictedNow more.

The reason is that it is possible that "evicted=false", but the workload is actually evicted, by another controller. Thus we have the check ` workload.IsEvicted(&wl) `.